smarty/forcefield.py

```diff
@@ -1462,22 +1596,40 @@ def labelForce(self, oemol, verbose=False, **kwargs):
 
         # Iterate over all defined torsion types, allowing later matches to override earlier ones.
         torsions = ValenceDict()
-        for torsion in self._torsiontypes:
+        for torsion in self._propertorsiontypes:
             for atom_indices in getSMIRKSMatches_OEMol(oemol, torsion.smirks, aromaticity_model = self.ff._aromaticity_model):
                 torsions[atom_indices] = torsion
+        # Handle impropers in similar manner
+        impropers = ImproperDict()
+        for improper in self._impropertorsiontypes:
+            for atom_indices in getSMIRKSMatches_OEMol(oemol, improper.smirks, aromaticity_model = self.ff._aromaticity_model):
+                impropers[atom_indices] = improper
 
         if verbose:
             print('')
             print('PeriodicTorsionGenerator:')
             print('')
-            for torsion in self._torsiontypes:
+            for torsion in self._propertorsiontypes:
                 print('%64s : %8d matches' % (torsion.smirks, len(getSMIRKSMatches_OEMol(oemol, torsion.smirks, aromaticity_model = self.ff._aromaticity_model))))
             print('')
+            print('PeriodicTorsionGenerator Impropers:')
+            print('')
+            for improper in self._impropertorsiontypes:
+                print('%64s : %8d matches' % (improper.smirks, len(getSMIRKSMatches_OEMol(oemol, improper.smirks, aromaticity_model = self.ff._aromaticity_model))))
+            print('')
 
         # Add all torsions to the output list
         force_terms = []
         for (atom_indices, torsion) in torsions.items():
             force_terms.append( ([atom_indices[0], atom_indices[1], atom_indices[2], atom_indices[3]], torsion.pid, torsion.smirks) )
+        # Add all impropers to the output list
+        for (atom_indices, improper) in impropers.items():
```

Comment:
Same comment as above: You can use [`permutations`](https://docs.python.org/2/library/itertools.html#itertools.permutations).